metronome/core/src/io/iohk/metronome/core/messages/RPCTracker.scala

```diff
@@ -0,0 +1,89 @@
+package io.iohk.metronome.core.messages
+
+import cats.implicits._
+import cats.effect.{Concurrent, Timer, Sync}
+import cats.effect.concurrent.{Ref, Deferred}
+import java.util.UUID
+import scala.concurrent.duration.FiniteDuration
+import scala.reflect.ClassTag
+
+/** `RPCTracker` can be used to register outgoing requests and later
+  * match them up with incoming responses, thus turning the two independent
+  * messages into a `Kleisli[F, Request, Option[Response]]`, where a `None`
+  * result means the request timed out.
+  */
```

Comment:
I understand you mean that this is _equivalent_ to a Kleisli, but I still fail to see the similarity. Could you explain please?

Reply:
Sorry, that was a bit vague, it doesn't turn them into a `Request => F[Option[Response]]` per se, but it facilitates that pattern by a component that does network communication.

What I meant is that we have this `Network` interface where you have a stream of incoming messages, and you can send messages in a fashion similar to an Akka Actor, i.e. `send: Message => F[Unit]`. But in many places it would be useful if we could send a request and wait for the response. We can do that by adding a listener to the incoming stream and matching the `requestId`.

This `RPCTracker` allows us to do that: every `Request` has a `requestId`, which a `Response` has to echo back, and the tracker will register an outgoing request (has to be done before it's sent away), and later can handle the `Response` by completing the equivalent of a `Deferred`. It also handles timeouts, so we never wait forever. The `F[F[Option[Response]]]` returned by `submit` can be used to wait for the response, or the timeout.

So`RPCTracker` is used by the `SyncService` to present network communication as a Kleisli.

Reply:
Changed the wording and added a workflow description.